keps/sig-etcd/4578-server-feature-gate/README.md

```diff
@@ -0,0 +1,338 @@
+
+The `ServerConfig` struct will have a new `featuregate.FeatureGate`(immutable) field. `EtcdServer` would have an interface of `FeatureEnabled(key Feature) bool`, and it can be referenced throughout the server.
+
+```go
+type ServerConfig struct {
+  ...
+  // ServerFeatureGate server level feature gate
+  ServerFeatureGate featuregate.FeatureGate
+  ...
+}
+
+func (s *EtcdServer) FeatureEnabled(key Feature) bool
+```
+
+### Ways to Query Feature Gate State
+
+New Prometheus gauge metrics would be added to monitor and query if a feature is enabled for the server. 
+
+The metrics could look like:
+* `etcd_feature_enabled{name="featureName",stage="Alpha"} 1` if the feature is enabled.
+
+### Feature Stages
+
+We propose etcd features to follow the convention of [Kubernetes feature stages](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages). Generally, a feature can go through a lifecycle of Alpha → Beta → GA → Deprecated. Each feature should be associated with a Kubernetes Enhancement Proposal ([KEP](https://www.kubernetes.dev/resources/keps/)), with high level graduation criteria defined in the KEP.
+
+| Feature Stage | Properties | Minimum Graduation Criteria |
+| --- | --- | --- |
+| Alpha | Same as [Kubernetes feature stages - Alpha feature](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages): <ul><li>Disabled by default. </li><li>Might be buggy. Enabling the feature may expose bugs. </li><li>Support for feature may be dropped at any time without notice. </li><li>The API may change in incompatible ways in a later software release without notice. </li><li>Recommended for use only in short-lived testing clusters, due to increased risk of bugs and lack of long-term support.</li></ul> | Before moving a feature to Beta, it should have <ul><li> Full unit/integration/e2e/robustness test coverage.</li><li>Full performance benchmark/test if applicable.</li><li> No significant changes for at least 1 minor release.</li></ul> |
+| Beta | Same as [Kubernetes feature stages - Beta feature](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages) except etcd Beta feature does not allow incompatible schema change: <ul><li>Enabled by default. </li><li>The feature is well tested. Enabling the feature is considered safe.</li><li>Support for the overall feature will not be dropped, though details may change.</li><li>Recommended for only non-business-critical uses because of potential for discovering new hard-to-spot bugs through wider adoption.</li></ul> | Before moving a feature to GA, it should have <ul><li> Widespread usage.</li><li>No bug reported for at least 1 minor release.</li></ul> |
```

Comment:
Minor comment:
- For Alpha -> Beta, should add Criteria: `No unresolved major bugs`.
- For Beta -> GA, update criteria:  `No major bug reported for at least 1 minor release`

Note they are just minimal Graduation Criteria.